homeassistant/components/homekit_controller/config_flow.py

```diff
@@ -0,0 +1,263 @@
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import callback
+
+from .const import DOMAIN, KNOWN_DEVICES
+from .connection import get_bridge_information, get_accessory_name
+
+
+HOMEKIT_IGNORE = [
+    'BSB002',
+    'Home Assistant Bridge',
+    'TRADFRI gateway',
+]
+HOMEKIT_DIR = '.homekit'
+PAIRING_FILE = 'pairing.json'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def load_old_pairings(hass):
+    """Load any old pairings from on-disk json fragments."""
+    old_pairings = {}
+
+    data_dir = os.path.join(hass.config.path(), HOMEKIT_DIR)
+    pairing_file = os.path.join(data_dir, PAIRING_FILE)
+
+    # Find any pairings created with in HA 0.85 / 0.86
+    if os.path.exists(pairing_file):
+        with open(pairing_file) as pairing_file:
+            old_pairings.update(json.load(pairing_file))
+
+    # Find any pairings created in HA <= 0.84
+    if os.path.exists(data_dir):
+        for device in os.listdir(data_dir):
+            if not device.startswith('hk-'):
+                continue
+            alias = device[3:]
+            if alias in old_pairings:
+                continue
+            with open(os.path.join(data_dir, device)) as pairing_data_fp:
+                old_pairings[alias] = json.load(pairing_data_fp)
+
+    return old_pairings
+
+
+@callback
+def find_existing_host(hass, serial):
+    """Return a set of the configured hosts."""
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.data['AccessoryPairingID'] == serial:
+            return entry
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class HomekitControllerFlowHandler(config_entries.ConfigFlow):
+    """Handle a HomeKit config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the homekit_controller flow."""
+        self.model = None
+        self.hkid = None
+        self.devices = {}
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(self, user_input=None):
+        """Handle a flow start."""
+        import homekit
+
+        errors = {}
+
+        if user_input is not None and 'device' in user_input:
+            key = user_input['device']
+            if key in self.devices:
+                props = self.devices[key]['properties']
+                self.hkid = props['id']
+                self.model = props['md']
+                return await self.async_step_pair()
+
+            errors['device'] = 'unknown_device'
+
+        controller = homekit.Controller()
+        all_hosts = await self.hass.async_add_executor_job(
+            controller.discover, 5
+        )
+
+        self.devices = {}
+        for host in all_hosts:
+            status_flags = int(host['properties']['sf'])
+            paired = not status_flags & 0x01
```

Comment:
I have raised https://github.com/jlusiardi/homekit_python/issues/117.

Reply:
It may be a while before the upstream ticket is done and in a release. Ideally I'd address this point in a follow up PR after that release (there have already been 2 HA contributions that have conflicted with this set of PR's so i'm keep to avoid further bit rot).